### ISSUE TITLE
Cannot find module 'vue' when not providing it in a plugin project

### ISSUE THREAD

**Issue by kleinfreund** [vuejs/vue-test-utils-next#257]
A Vue.js plugin project with no dependency on the vue package doesn’t seem to be able to declare vue purely as a peer dependecy. When doing so, running the unit tests written using vue-test-utils (and vue-jest) results in the following error:

```
Cannot find module 'vue' from 'node_modules/@vue/test-utils/dist/vue-test-utils.cjs.js'

Require stack:
  node_modules/@vue/test-utils/dist/vue-test-utils.cjs.js
  src/ColorPicker.test.js
```

This is unexpected as vue-test-utils does include vue as a development dependency. However, it’s not being installed and is therefor absent from the node_modules directory. Perhaps I’m not fully understanding how exactly dependency resolution works for those declared under `devDependencies`?

A workaround for this issue is installing vue as a development dependency in the plugin, but that shouldn’t be necessary, I think.

---

One reason for why this is problematic for plugin authors is that, at least to my understanding, it prevents a local development workflow using `npm link` and `npm link <package-name>`: With the plugin having its own copy of the vue package, rendering the plugin within a consuming project results in rendering errors because now two separate vue packages are being used. In the vue code that selects the correct routine for handling different kinds of nodes ([runtime-core/src/**renderer.ts**#L475-L551](https://github.com/vuejs/vue-next/blob/master/packages/runtime-core/src/renderer.ts#L475-L551)), `Symbol` objects (as defined in [runtime-core/src/**vnode.ts**#L43-L51](https://github.com/vuejs/vue-next/blob/master/packages/runtime-core/src/vnode.ts#L43-L51)) are being used for comparison. This is where the problems start: The nodes from the plugin are compared with `Symbol`s in the consuming project’s vue package version that would be the same were the `Symbol`s created by the same script files. But they weren’t because both the consuming project and the plugin have their own vue package bundled. And so rendering of the components fails when it shouldn’t.

**Comment by lmiller1990** [comment]
I am not clear on what you are proposing. What change are you suggesting? The title says "Add vue as development dependency", but we already have it under `devDependencies`.

What exactly are you suggesting?

**Comment by kleinfreund** [comment]
Oh, well that’s embarrassing. I did not see it was there. I think I just assumed it wasn’t after spotting that vue was listed as a peer dependency.

What I’m currently running into is that upon removing the `vue` and `@vue/compiler-sfc` packages from my project, its unit tests start to fail with the following error:

```
Cannot find module 'vue' from 'node_modules/@vue/test-utils/dist/vue-test-utils.cjs.js'

Require stack:
  node_modules/@vue/test-utils/dist/vue-test-utils.cjs.js
  src/ColorPicker.test.js
```

And indeed, the vue package is no longer to be found in my project’s node_modules directory. I’m not sure if I’m doing something wrong, but given that the test utils package does, in fact, list vue as a development dependency, I would expect this error to not appear.

**Comment by afontcu** [comment]
Hi! I'm a bit lost, too

> upon removing the vue and @vue/compiler-sfc packages from my project, its unit tests start to fail with the following error

Vue is listed as a peerDep, so if you remove it from your project VTU will fail. You need to manually install peerDeps (at least until npm 7 is released!).

Also, as far as I can tell devDependencies are not installed transitively. So if Vue Test Utils lists Vue as a dependency, installing Vue Test Utils won't get you Vue.

**Comment by kleinfreund** [comment]
I see. So since I’m using `npm link` to test my plugin in a different project, the installed dev dependency of my plugin is being found upon resolving the package dependencies which wouldn’t happen in a regular `npm install` scenario because that does _not_ install the development dependency. In other words: having it as a development dependency is fine, in general, it just, unfortunately, breaks that aspect of testing it via `npm link`.

Of course none of that is really a fault of vue or vue-test-utils so I’m closing this issue. Thank you for your explanations and patience.

**Comment by afontcu** [comment]
Noo prob! Glad you end up sorting it out :)

**Comment by kleinfreund** [comment]
I just wanted to leave a note for anyone who might find this issue on why developing plugins for Vue.js 3 is currently cumbersome.

First, some observations:

- 1. vue-test-utils declares the vue package as a peer dependency.
- 2. Plugins whose unit tests rely on vue-test-utils need to install the vue package as a development dependency because of 1.
- 3. Plugins shouldn’t have a production dependency on the vue package because their consuming project must provide the vue package in any case.
- 4. Vue projects have a production dependency on the vue package.

Scenario: We have a Vue.js 3 project which uses a Vue plugin. This could be a demo website for the plugin.

In such a project, we have both the vue package and the plugin package installed. To test such a setup locally, one can usually use `npm install` followed by `npm link` in the plugin package’s local directory to make the package available on the user’s machine. In the project’s local directory, one would then run `npm install` and `npm link plugin-package-name`.

Now, when starting a local development server (perhaps with `vue-cli-service serve` when building a vue-cli project), rendering a component exposed by the plugin package **will most likely fail** because two separate instances of the vue package will be in use at this point. The project will use its copy of vue while the plugin project will use _a different copy from its `node_modules` directory_. This is because vue is a development dependency of the plugin project (see observation 2) and as such, vue was installed previously when running `npm install` before linking the plugin package. Due to this, vue’s rendering logic will fail when comparing to value-equal `Symbol`s that come from different realms and are therefor _not_ equal.

In other words, either of the following facts _break_ a local plugin development workflow using `npm link` in Vue.js 3:

- Vue.js using `Symbol`s within its rendering logic.
- Plugin projects being required to install vue as a development dependency.

**Comment by lmiller1990** [comment]
Correct me if I am wrong but it seems like you are relying on another *separate* project to test your plugin. Wouldn't the ideal scenario be to develop your plugin in a separate package with `vue` installed as a dev dependency, and perhaps have a simple test app there? 

It seems like the `npm link` part of your workflow is the main problem :thinking:

**Comment by kleinfreund** [comment]
In my case, I have a plugin and a demo website:

- [github.com/kleinfreund/**vue-accessible-color-picker**](https://github.com/kleinfreund/vue-accessible-color-picker)
- [github.com/kleinfreund/**vue-accessible-color-picker-website**](https://github.com/kleinfreund/vue-accessible-color-picker-website)

The plugin and all its tests and publishing logic are within the package repository. The demo website is really just that: a small Vue.js project making use of the plugin (which provides a component) for demonstration purposes. So when you write “you are relying on another _separate_ project to test your plugin” you are somewhat correct: I _do_ want to verify that the plugin can be consumed according to its documentation by using it in a Vue.js project as an actual user would do. I think it’s reasonable to have this be covered by a separate project. I don’t want to have a sub project within my plugin repository with its own set of dependencies.

I think all I’m trying to say is that it would be _nice_ if one could develop a Vue.js plugin similar to what I did (i.e. vue is a development dependency) and test its behavior using `npm link` because that would be simple and allow people to have separate projects for a plugin consumer and the plugin itself.

I also wonder if vue’s core logic _requires_ the mentioned `Symbol`-based checks to actually use `Symbol`s. If checks based on plain string variables would do, then this problem would be indirectly addressed, too. Of course if the use of `Symbol`s in this context has some advantages, that doesn’t hold.

**Comment by lmiller1990** [comment]
I think a `Symbol` makes sense here - I understand what you are doing now, but I am sure it is not a super common workflow. I think one solution to this is using something like `lerna` where you can manage multiple projects in one, all sharing dependencies.

Also, if you just mark the test project's deps as dev dependencies, (I personally) do not see what is wrong with having them in the same repo as the plugin source - I actually prefer this, since then I only need one repo for dev.

Anyway, if there is something I can do here to help, let me know. Good luck 👍  share a link to your plugin!
